cms/lib/xblock/container_upstream_sync.py

```diff
@@ -0,0 +1,202 @@
+        try:
+            lib_meta = get_container(upstream_key)
+        except ContentLibraryContainerNotFound as exc:
+            raise BadUpstream(_("Linked library item was not found in the system")) from exc
+        return cls(
+            upstream_ref=downstream.upstream,
+            version_synced=downstream.upstream_version,
+            version_available=(lib_meta.published_version_num if lib_meta else None),
+            version_declined=downstream.upstream_version_declined,
+            error_message=None,
+        )
+
+    @classmethod
+    def try_get_for_container(cls, downstream: XBlock) -> t.Self:
+        """
+        Same as `get_for_container`, but upon failure, sets `.error_message` instead of raising an exception.
+        """
+        try:
+            return cls.get_for_container(downstream)
+        except UpstreamLinkException as exc:
+            logger.exception(
+                "Tried to inspect an unsupported, broken, or missing downstream->upstream link: '%s'->'%s'",
+                downstream.usage_key,
+                downstream.upstream,
+            )
+            return cls(
+                upstream_ref=downstream.upstream,
+                version_synced=downstream.upstream_version,
+                version_available=None,
+                version_declined=None,
+                error_message=str(exc),
+            )
+
+
+def _get_library_container_url(container_key: LibraryContainerLocator):
+    """
+    Gets authoring url for given library_key.
+    """
+    library_url = None
+    if mfe_base_url := settings.COURSE_AUTHORING_MICROFRONTEND_URL:  # type: ignore
+        library_key = container_key.context_key
+        library_url = f'{mfe_base_url}/library/{library_key}/units?container_key={container_key}'
+    return library_url
+
+
+class ContainerUpstreamSyncManager(BaseUpstreamSyncManager):
+    def __init__(self, downstream: XBlock, user: User, upstream: XBlock | None = None) -> None:
+        super().__init__(downstream, user, upstream)
+        if not isinstance(self.upstream_key, LibraryContainerLocator):
+            raise BadUpstream('Invalid upstream key')
+        self.link = ContainerUpstreamLink.get_for_container(downstream)
+        if not upstream:
+            self.upstream = self._load_upstream_link_and_container_block()
+        self.syncable_field_names: set[str] = self._get_synchronizable_fields()
+        self.new_children_blocks: list[XBlock] = []
+
+    def _get_synchronizable_fields(self) -> set[str]:
+        """
+        The syncable fields are the ones which are content- or settings-scoped AND are defined on both (up,down)stream.
+        """
+        return set.intersection(*[
+            set(
+                field_name
+                for (field_name, field) in block.__class__.fields.items()
+                if field.scope in [Scope.settings, Scope.content]
+            )
+            for block in [self.upstream, self.downstream]
+        ])
```

Comment:
For Teak, containers probably won't have any fields other than `display_name`, so that's likely the only one necessary to sync. Also container fields don't have scopes.

If we manage to implement https://github.com/openedx/frontend-app-authoring/issues/1640 soon, we may have two actual fields on units, but we can add the field syncing then.